### PR TITLE
Fingerbank v2 integration + collector

### PR DESCRIPTION
**Pull request by julsemaan**
## NOT READY

# Missing

- [x] Upgrade notes for the API key which is now mandatory for device profiling
- [x] Upgrade notes for deleting pf_fingerbank
- [x] Have the collector persist its data on disk for restoration after a restart
- [x] Have the collector work in a cluster mode (see the Inverse internal wiki)
- [ ] Potentially populate the collector from the data that is stored in the PacketFence database 
- [ ] Before v8 release, sync fingerbank.inverse.ca accounts created after V2 launch on api.fingerbank.org
- [x] Display account information and link to api.fingerbank.org account directly in PacketFence

# Description
Integrates PacketFence with the latest version of the Fingerbank API + integrates with the collector

# Impacts
Device profiling

# NEW Package(s) required
fingerbank-4.0.0
fingerbank-collector-1.0.0

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Now integrating with the second version of Fingerbank (api.fingerbank.org) for improved device profiling
* Now using the Fingerbank collector to perform device profiling using Fingerbank. Provides improved TCP fingerprinting, mDNS profiling, UPnP based profiling, and much more

# UPGRADE file entries

## Configuring a Fingerbank API key

Having a Fingerbank API key is now mandatory in order to be able to perform device profiling in PacketFence (i.e. detecting the operating system of the endpoints on your network).

In order to make sure you have a Fingerbank API key configured, execute the following:
```
# grep api_key /usr/local/fingerbank/conf/fingerbank.conf
```

The command above should output your Fingerbank API key. If it provides no output, then register a Fingerbank account here: https://api.fingerbank.org/users/register. After registering your account, execute the following command:

```
# export K="PUT_YOUR_API_KEY_HERE" 
# perl -Ilib -Mfingerbank::Config -e "fingerbank::Config::write_config({upstream => {api_key => '$K'}})"
```

Fingerbank API requests with free accounts are limited to 300 requests per hour. This should be enough for most small scale deployments. E-mail alerts will be sent when reaching that limit. If you need to obtain an upgraded Fingerbank account, contact info@inverse.ca. Inverse customers are entitled to upgraded accounts, so if you are already an Inverse customer, contact Inverse to obtain a corporate account or to upgrade your existing API key.

## Deleting the Fingerbank MySQL database

The Fingerbank library isn't using MySQL for storing the upstream database anymore. although it won't do any harm to leave the database in place, its suggested to remove it.

In order to remove it, connect to MySQL, then list your databases:
```
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| pf                 |
| pf_fingerbank      |
| pf_graphite        |
+--------------------+
```

The Fingerbank database should be named pf_fingerbank or the name of your current database name suffixed by `_fingerbank`. In this example, it is `pf_fingerbank`.

Now in order to remove it, perform the following query:
```
mysql> drop database pf_fingerbank;
```

### CODE REVIEW

**Comment by julsemaan** [comment]
PR is ready for review.

Stuff that would need to be looked at post-merge:
- [ ] Potentially populate the collector from the data that is stored in the PacketFence database 
- [ ] Before v8 release, sync fingerbank.inverse.ca accounts created after V2 launch on api.fingerbank.org
- [ ] Consider not restoring the IP database of the collector if its older than X time
- [ ] Handle re-sending sync packets in the collector (for which we got no ACK)
- [ ] Forget the attributes of an endpoint when detecting a change in the DHCP vendor
- [ ] Add support to synchronize the deletion of an endpoint (needed for point above)

**Comment by extrafu** [comment]
Excellent work. @fdurand please review and merge.

**Comment by julsemaan** [comment]
@fdurand 

Please also take a look at fingerbank/perl-client#33 and https://git.inverse.ca/inverse/fingerbank-collector/pulls/19

File: html/pfappserver/lib/pfappserver/Form/Config/Fingerbank/Settings.pm [hunk 1]
```diff
@@ -13,6 +13,7 @@ Form definition to modify Fingerbank configuration
 use HTML::FormHandler::Moose;
 
 use fingerbank::Config;
+use pf::log;
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/Fingerbank/Settings.pm]: Looks that it's no used

File: lib/pf/condition/fingerbank/device_is_a.pm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    return fingerbank::Model::Device->is_a($device, $self->value);
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2017 Inverse inc.
```
Comment (fdurand) [inline lib/pf/condition/fingerbank/device_is_a.pm]: 2018

File: lib/pf/services/manager/fingerbank_collector.pm [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    return $TRUE;
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2017 Inverse inc.
```
Comment (fdurand) [inline lib/pf/services/manager/fingerbank_collector.pm]: 2018

**Review by fdurand — COMMENTED** [review]
As i can see it's just details to change.
Fix it and i will merge.

**Comment by julsemaan** [comment]
Adjusted based on comments

**Pull request merged**